### PR TITLE
Add Assert.AreEquivalent for deep structural comparison

### LINKED ISSUE #4776: Add Assert.AreEquivalent to perform deep comparison of objects
## Summary

Add an `Assert.AreEquivalent` method that performs a deep compare on two objects, similar to for example the xUnit `Assert.Equivalent` method or FluentAssertion's `actual.Should().BeEquivalentTo(expected)`.

## Background and Motivation

I'm trying to perform a deep compare of two objects based on the values of the properties, with the option to exclude certain properties. 

I was using `FluentAssertions` for this, but the latest version got a license change where it's now a paid license for commercial use. The only reason I was using this library is because MSTest doesn't provide a deep comparison option. Instead of having to switch to a different library, it would be useful to have this feature in MSTest itself like xUnit. Reducing the number of dependencies and having better feature parity with xUnit.

## Proposed Feature

You can pass two objects into `Assert.AreEquivalent(expected, actual)` and it will perform a comparison of all (public) fields and properties. It will raise an error if the value of a field/property doesn't match. Using a parameter, you can specify that certain properties should be ignored.

It would also be nice if it could perform a deep compare of the items in two collections. E.g. comparing the first item of each collection using the `Assert.AreEquivalent` method.

### PR DESCRIPTION
**Pull request by Evangelink**
## What

Adds `Assert.AreEquivalent<T>` and `Assert.AreNotEquivalent<T>` for deep structural equality comparison of object graphs, addressing the long-standing request in #4776.

```csharp
// Pass: deep structural compare of two POCOs
Assert.AreEquivalent(new Order("o-1", new Address("street", "city")),
                     new Order("o-1", new Address("street", "city")));

// Fail with: Mismatch at 'ShippingAddress.City': values are not equal.
//            expected: "city"
//            actual:   "different"
Assert.AreEquivalent(new Order("o-1", new Address("street", "city")),
                     new Order("o-1", new Address("street", "different")));
```

## Why

MSTest currently has:
- `Assert.AreEqual` — reference / `Equals` / `IEqualityComparer<T>`-based, useless for POCOs with default `object.Equals`.
- `CollectionAssert.AreEqual` — ordered element-by-element.
- `CollectionAssert.AreEquivalent` — set-equivalence.

There's no way to assert that two arbitrary object graphs are structurally equal. Users have been reaching for FluentAssertions' `BeEquivalentTo`, but its license recently changed to commercial. xUnit has had `Assert.Equivalent` for a while.

## Design

API shape follows xUnit's `Assert.Equivalent` more than FluentAssertions' `BeEquivalentTo`, in line with MSTest's preference for a small focused API surface that can grow without breaking changes:

- 8 new public overloads (4 each for `AreEquivalent` and `AreNotEquivalent`), with optional `bool strict` and standard `CallerArgumentExpression` plumbing. No options bag for now — it can be layered on later as a pure addition.
- **Comparison rules** (full list in the XML `<remarks>`):
  - Reference-equal values (and both `null`) are equivalent.
  - **Primitive-like** types (numerics, `string`, `bool`, `char`, enums, `DateTime`, `DateTimeOffset`, `TimeSpan`, `Guid`, `Uri`, `Type`, `Version`, plus `DateOnly`/`TimeOnly`/`Half`/`Int128`/`UInt128` when present) → `Equals`.
  - **Hybrid `IEquatable<T>` shortcut**: when the *static* declared type implements `IEquatable<self>`, dispatch via the interface method (so explicit interface implementations work too). Plain `object.Equals` overrides are intentionally ignored — recursion proceeds.
  - **Dictionaries** (`IDictionary`, `IDictionary<,>`, or `IReadOnlyDictionary<,>`) are compared by key set with values recursive. Lookups are routed through the source dictionary so its `IEqualityComparer<TKey>` is preserved (e.g., `StringComparer.OrdinalIgnoreCase`).
  - **Other enumerables** (excluding `string`) are compared element-by-element in iteration order. Element type is taken from the declared `IEnumerable<T>` so the `IEquatable<T>` shortcut still applies for nested values. (For order-insensitive comparison use `CollectionAssert.AreEquivalent` on the collection itself.)
  - **Other reference types** are compared by recursing into all public instance properties (with a public getter, no index parameters) and public instance fields, sorted by name for stable diagnostics.
  - **Cycles + topology**: tracked via persistent bidirectional ref-equality maps. Re-entering a known pair short-circuits as match. Re-entering with a different counterpart on either side fails as a topology mismatch.
  - **User exceptions**: thrown by `IEquatable<T>.Equals`, member getters, or dictionary access (`TryGetValue`/`ContainsKey`/enumeration) are caught and surfaced as a structured failure rather than escaping `AssertFailedException`.
- **`strict: true`** additionally rejects extra public members on `actual` (when runtime types differ) and any extra dictionary keys on `actual`.

Failure messages use the existing structured assertion infrastructure (RFC 012): a top-line summary, a `Mismatch at '<dotted-path>': <reason>` locator (e.g. `Order.Items[2].Price` or `["x-trace-id"]`), and `expected:` / `actual:` evidence rendered through `AssertionValueRenderer`. `AssertFailedException.ExpectedText`/`ActualText` are populated for any mismatch that can be rendered as two values.

### Naming considered
`Assert.AreEquivalent` was deliberately chosen over `AreDeeplyEqual` / `AreStructurallyEqual` to match xUnit and the wording in #4776, even though `CollectionAssert.AreEquivalent` already means set-equivalence. The two APIs live on different types and accept different parameter types, so there's no compile-time conflict.

## Implementation

Two new files:

- `src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.cs` — public surface, structured-message reporting helpers. Mirrors the style of `Assert.AreEqual.cs` / `Assert.AreSame.cs`.
- `src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs` — internal `EquivalenceComparer` (recursive walker), `MemberLookup` (cached per type, with both sorted array and O(1) name-keyed dict), `DictionaryView` / `NonGenericDictionaryView` / `GenericDictionaryView` / `GenericDictionaryAccessors` (cached reflection that defers lookups to the source dictionary), `EquivalenceMismatch` (structured failure record).

15 new strings added to `FrameworkMessages.resx`; sibling XLFs regenerated via `dotnet msbuild ... /t:UpdateXlf`. 8 new entries in `PublicAPI.Unshipped.txt`.

## Tests

`test/UnitTests/TestFramework.UnitTests/Assertions/AssertTests.AreEquivalentTests.cs` — 62 tests, TestContainer + AwesomeAssertions, covering: primitives/null/strings/dates/enums/Guid; identical and different POCOs (with dotted-path failure locators); public fields; anonymous types; ignored private fields; collections (length/element/order); nested collections; dictionaries (equal, missing key, different value, extra key non-strict vs strict, non-string keys, case-insensitive comparer respected, custom `IReadOnlyDictionary<,>`-only types); `IEquatable<T>` (regular, explicit interface implementation, value-type, throws); plain `Equals` override is ignored; self-cycles and cross-cycles via property and field; topology mismatch detection (shared-vs-distinct); value-type topology skip; member-getter throws on expected and actual sides; dictionary-access throws; strict-mode extra-member rendering (alphabetical); user message and call-site expression rendering; plus broad `AreNotEquivalent` mirroring (collections, dictionaries, cycles, IEquatable, call-site expression).

## Verification

- `dotnet build src/TestFramework/TestFramework -c Debug /p:TreatWarningsAsErrors=true` — clean across `netstandard2.0`, `net462`, `net8.0`, `net9.0`.
- `dotnet build test/UnitTests/TestFramework.UnitTests -c Debug /p:TreatWarningsAsErrors=true` — clean across all TFMs (incl. WinUI).
- 62 `AreEquivalent` tests pass on **net9.0** and **net48**.
- Full `TestFramework.UnitTests` suite (954 tests) green.

## Out of scope (follow-ups)

- `EquivalenceOptions` bag (exclude paths, max-depth, ignore-collection-order, configure key/value rendering). Can be added without breaking changes.
- Analyzer suggesting `Assert.AreEquivalent` over `Assert.AreEqual` when the latter is called on POCO types.

Closes #4776.

### CODE REVIEW

File: src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs [hunk 1]
```diff
@@ -0,0 +1,1004 @@
...
+            List<object?> list = source is ICollection collection
+                ? new List<object?>(collection.Count)
+                : [];
+#pragma warning restore IDE0028
+            foreach (object? item in source)
+            {
+                list.Add(item);
+            }
+
+            return list;
```
Comment (Evangelink) [inline src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs]: **[MAJOR] Defensive Coding at Boundaries**

`ToList` iterates a user-provided `IEnumerable` with no exception handling. If a custom `IEnumerable` throws from `GetEnumerator()`, `MoveNext()`, or `Current`, the raw exception propagates through `CompareEnumerables` and out of the public API as an unhandled exception — not a structured `AssertFailedException`.

**Concrete scenario:**
```csharp
class ThrowingList : IEnumerable<int> {
    public IEnumerator<int> GetEnumerator() => throw new InvalidOperationException("boom");
    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
}
Assert.AreEquivalent(new ThrowingList(), new ThrowingList());
// → throws raw InvalidOperationException, NOT AssertFailedException
```

This is inconsistent with:
- `ForEachEntry` (lines 178–225), which wraps dictionary enumeration exceptions with structured failure messages.
- `CompareMembers`, which catches member-getter exceptions and returns `MemberAccessFailure`.
- The design spec: *"thrown by... dictionary access (... enumeration) are caught and surfaced as a structured failure"*.

**Recommendation:** Wrap both `ToList` calls in `CompareEnumerables` with try/catch (similar to `ForEachEntry`), or add try/catch inside `ToList` that converts any user-thrown exception to `EquivalenceMismatch.EnumerationFailure(path, isExpected, ex)`. A new `EnumerationFailure` factory method on `EquivalenceMismatch` (analogous to `MemberAccessFailure`) would keep the pattern consistent.

**Review by Evangelink — COMMENTED** [review]
| # | Dimension | Verdict |
|---|-----------|---------|
| 8 | Defensive Coding at Boundaries | 🟡 1 MAJOR |

✅ 20/21 dimensions clean.

- [ ] Defensive Coding — `ToList` doesn't catch user-thrown enumeration exceptions; raw exceptions escape the public API instead of becoming a structured `AssertFailedException`

---

**Overall**: This is a well-designed, well-implemented feature. The implementation is thorough — reflection caching, cycle/topology detection, per-entry exception wrapping in dictionaries, IEquatable shortcut, proper `PublicAPI.Unshipped.txt` entries, localization via `.resx` only, 62 tests covering edge cases. The single finding is that `IEnumerable` enumeration exceptions are not caught while dictionary enumeration exceptions (via `ForEachEntry`) and member-getter exceptions (via `CompareMembers`) are — a straightforward oversight to fix.

> Generated by [Expert Code Review (on open)](https://github.com/microsoft/testfx/actions/runs/25928209616/agentic_workflow) for issue #8266 · ● 18.3M

File: src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.cs [hunk 2]
```diff
@@ -0,0 +1,294 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestTools.UnitTesting;
+
+public sealed partial class Assert
```
Comment (Evangelink) [inline src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.cs:6]: Let's make sure this is following RFC 12

File: test/UnitTests/TestFramework.UnitTests/Assertions/AssertTests.AreEquivalentTests.cs [hunk 3]
```diff
@@ -0,0 +1,914 @@
...
+        Assert.AreEquivalent(3.14, 3.14);
+        Assert.AreEquivalent(true, true);
+        Assert.AreEquivalent('z', 'z');
+    }
+
+    public void AreEquivalent_DifferentInts_Fails()
+    {
+        Action act = () => Assert.AreEquivalent(1, 2);
+        AssertFailedException ex = act.Should().Throw<AssertFailedException>().Which;
+        ex.Message.Should().Contain("Mismatch at '<root>'");
```
Comment (Evangelink) [inline test/UnitTests/TestFramework.UnitTests/Assertions/AssertTests.AreEquivalentTests.cs:66]: Always do full match using """ instead of contains or Environment.NewLine checks.

File: src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.cs [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+        EquivalenceComparer comparer = new(strict);
+        EquivalenceMismatch? mismatch = comparer.Compare(notExpected, actual);
+        if (mismatch is not null)
+        {
+            return;
+        }
+
+        ReportAssertAreNotEquivalentFailed(notExpected, actual, strict, message, notExpectedExpression, actualExpression);
+    }
+
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.cs:256]: Fixed in 19b9599ed: comparison failures now carry an IsComparisonFailure flag, Assert.AreNotEquivalent rethrows them as structured AssertFailedExceptions, and coverage now includes throwing IEquatable/enumerable scenarios.

File: src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs [hunk 5]
```diff
@@ -0,0 +1,1004 @@
...
+                if (_actualToExpected.TryGetValue(actual, out object? mappedExpected))
+                {
+                    return ReferenceEquals(mappedExpected, expected)
+                        ? null
+                        : EquivalenceMismatch.TopologyMismatch(path);
+                }
+
+                _expectedToActual[expected] = actual;
+                _actualToExpected[actual] = expected;
+            }
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs:123]: Fixed in 19b9599ed: I documented the fail-fast invariant around topology tracking so it is clear why the recorded pair is intentionally not unwound on a failed sub-compare.

File: src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs [hunk 6]
```diff
@@ -0,0 +1,1004 @@
...
+
+            if (expected is IEnumerable expectedEnum && actual is IEnumerable actualEnum)
+            {
+                Type elementDeclaredType = GetEnumerableElementType(declaredType, expectedRuntimeType, actualRuntimeType);
+                return CompareEnumerables(expectedEnum, actualEnum, elementDeclaredType, path);
+            }
+
+            return CompareMembers(expected, actual, expectedRuntimeType, actualRuntimeType, path);
+        }
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs]: Fixed in 19b9599ed: added a bounded recursion depth (256) with a structured max-depth mismatch instead of risking StackOverflowException, plus a deep-graph regression test.

File: src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs [hunk 7]
```diff
@@ -0,0 +1,1004 @@
...
+            => parent.Length == 0 ? name : $"{parent}.{name}";
+
+        private static string AppendIndex(string parent, int index)
+            => $"{parent}[{index.ToString(CultureInfo.InvariantCulture)}]";
+
+        private static string AppendDictionaryKey(string parent, object key)
+        {
+            string rendered = key is string s
+                ? $"\"{s}\""
+                : Convert.ToString(key, CultureInfo.InvariantCulture) ?? key.GetType().Name;
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs]: Fixed in 19b9599ed: dictionary key paths now route through AssertionValueRenderer.RenderValue, so the path and key diagnostics stay consistent. Added coverage for escaped string keys.

File: src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs [hunk 8]
```diff
@@ -0,0 +1,1004 @@
...
+            Type expectedRuntimeType = expected.GetType();
+            Type actualRuntimeType = actual.GetType();
+
+            // Primitive-ish types: trust Equals.
+            if (IsPrimitiveLike(expectedRuntimeType) || IsPrimitiveLike(actualRuntimeType))
+            {
+                return Equals(expected, actual)
+                    ? null
+                    : EquivalenceMismatch.ValueMismatch(path, expected, actual);
+            }
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs:77]: Fixed in 19b9599ed: primitive-like short-circuiting now reports a type mismatch when runtime types differ, so boxed object comparisons do not degrade to a generic value mismatch. Added regression coverage.

File: src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.cs [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+    ///     <see cref="IEqualityComparer{T}"/> for keys is preserved.
+    ///     </description>
+    ///   </item>
+    ///   <item>
+    ///     <description>
+    ///     Other enumerables (excluding <see cref="string"/>) are compared element-by-element in the
+    ///     iteration order. Comparison is <b>order-sensitive</b>; if order is irrelevant, use
+    ///     <see cref="CollectionAssert.AreEquivalent(System.Collections.ICollection?, System.Collections.ICollection?)"/>
+    ///     on the collection itself.
+    ///     </description>
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.cs:87]: Fixed in 19b9599ed: expanded the AreEquivalent XML remarks to explicitly contrast this deep, order-sensitive assertion with CollectionAssert.AreEquivalent.

File: src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs [hunk 10]
```diff
@@ -0,0 +1,1004 @@
...
+                {
+                    if (!expectedMembers.ByName.ContainsKey(am.Name))
+                    {
+                        (extras ??= []).Add(am.Name);
+                    }
+                }
+
+                if (extras is { Count: > 0 })
+                {
+                    extras.Sort(StringComparer.Ordinal);
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs]: Fixed in 19b9599ed: removed the redundant extras.Sort(...) call because MemberLookup.Sorted already guarantees deterministic ordinal member ordering.

File: src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs [hunk 11]
```diff
@@ -0,0 +1,1004 @@
...
+
+                // DateOnly / TimeOnly / Half / Int128 exist only on newer TFMs; match by full name to avoid #if.
+                string? fullName = t.FullName;
+                if (fullName is "System.DateOnly" or "System.TimeOnly" or "System.Half" or "System.Int128" or "System.UInt128")
+                {
+                    return true;
+                }
+
+                // Compare reflection types as primitive (they all derive from System.Type).
+                return typeof(Type).IsAssignableFrom(t);
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs]: Fixed in 19b9599ed: dropped the broad System.Type subclass fallback from IsPrimitiveLike and handled Type values explicitly in Compare, so boxed Type instances still use equality without classifying every subclass as primitive-like.

File: src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs (hunk 1 above)
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs]: Fixed in 19b9599ed: enumerable materialization is now wrapped in structured exception handling via EnumerationFailure, and both AreEquivalent/AreNotEquivalent now fail with AssertFailedException instead of leaking or swallowing user exceptions. Added regression coverage.

File: src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.cs (hunk 2 above)
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.cs:6]: Checked against RFC 012 in 19b9599ed: the new comparison-failure and depth-limit paths now flow through StructuredAssertionMessage, and I added a regression test that asserts the AreEquivalent failure layout still matches the RFC structure.

File: test/UnitTests/TestFramework.UnitTests/Assertions/AssertTests.AreEquivalentTests.cs (hunk 3 above)
Comment (Evangelink) [reply test/UnitTests/TestFramework.UnitTests/Assertions/AssertTests.AreEquivalentTests.cs:66]: Fixed in 8f73a91fa: switched the RFC 012 regression test to a full raw-string message match instead of piecemeal Contains checks.

File: src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.cs [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+        {
+            ReportAssertAreNotEquivalentFailed(notExpected, actual, strict, message, notExpectedExpression, actualExpression);
+        }
+
+        if (mismatch.IsComparisonFailure)
+        {
+            ReportAssertAreNotEquivalentComparisonFailed(mismatch, strict, message, notExpectedExpression, actualExpression);
+        }
+
+        return;
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.cs]: Fixed in d1910d06a4c5458637358a6b7765dd0aa8f91d12 by rewriting the flow as if/lse if and removing the trailing 
eturn, so the null/non-null cases are explicit without depending on [DoesNotReturn] for readability.

File: src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs [hunk 13]
```diff
@@ -0,0 +1,1072 @@
...
+            for (int i = 0; i < expectedItems.Count; i++)
+            {
+                EquivalenceMismatch? nested = Compare(expectedItems[i], actualItems[i], elementDeclaredType, AppendIndex(path, i), depth + 1);
+                if (nested is not null)
+                {
+                    return nested;
+                }
+            }
+
+            return null;
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs]: Fixed in d1910d06a4c5458637358a6b7765dd0aa8f91d12 by reworking CompareEnumerables to walk both enumerators in lockstep instead of materializing them into lists. It now fails fast on the first mismatching element, still reports length mismatches by draining only the longer tail, and keeps GetEnumerator/MoveNext/Current guarded with the same enumeration-failure handling. I also added a unit test that would fail if enumeration continued past the mismatch index.

File: src/TestFramework/TestFramework/Resources/FrameworkMessages.resx [hunk 14]
```diff
@@ -159,6 +159,83 @@
...
+  </data>
+  <data name="AreEquivalentMismatchActualEnumerationThrew" xml:space="preserve">
+    <value>enumerating the actual collection threw {0}: {1}.</value>
+  </data>
+  <data name="AreEquivalentMismatchExpectedMemberThrew" xml:space="preserve">
+    <value>reading the expected member threw {0}: {1}.</value>
+  </data>
+  <data name="AreEquivalentMismatchActualMemberThrew" xml:space="preserve">
+    <value>reading the actual member threw {0}: {1}.</value>
+  </data>
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/TestFramework/TestFramework/Resources/FrameworkMessages.resx:235]: Fixed in d1910d06a4c5458637358a6b7765dd0aa8f91d12 by reordering the new FrameworkMessages.resx entries back into alphabetical groups: the AreEquivalent* block now sits before the AreNot* entries, and the AreNotEquivalent* entries follow AreNotEqualDeltaFailMsg. I also ran the UpdateXlf target afterward.

File: src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs [hunk 15]
```diff
@@ -0,0 +1,1072 @@
...
+            {
+                return null;
+            }
+
+            if (expected is null || actual is null)
+            {
+                return EquivalenceMismatch.NullMismatch(path, expected, actual);
+            }
+
+            if (depth > MaxComparisonDepth)
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.AreEquivalent.Comparer.cs]: Fixed in d1910d06a4c5458637358a6b7765dd0aa8f91d12 by changing the depth guard to depth >= MaxComparisonDepth so the implementation matches the user-facing 256 limit. I also added coverage for the boundary: depth 256 passes, and the next level fails.
